File: host/Sources/VibeWireHost/Claude/ChannelBridge.swift
import Foundation

/// Talks to the VibeWire channel MCP server, which Claude Code spawns inside a
/// session that is already running.
///
/// This is the difference between starting a conversation from the phone and
/// joining the one open in a terminal. `ClaudeBridge` drives its own
/// `claude --print` process, which is a separate conversation by construction —
/// two `claude` processes cannot share one. A channel is the supported way in:
/// Claude Code owns the subprocess, and what this server emits is injected
/// straight into the live session.
///
/// The channel listens on loopback and is gated on a token, so this is the only
/// thing on the machine that can reach it.
actor ChannelBridge {
    typealias Emit = @Sendable ([String: Any]) -> Void

    private let port: Int
    private let session: URLSession
    private var emit: Emit?
    private var streamTask: Task<Void, Never>?
    private(set) var isAttached = false

    /// The id of the last message sent, so Claude's reply can be matched to it.
    private var lastChatId: String?

    init(port: Int = 8790) {
        self.port = port
        let configuration = URLSessionConfiguration.default
        // The event stream is deliberately open-ended.
        configuration.timeoutIntervalForRequest = 3600
        self.session = URLSession(configuration: configuration)
    }

    private var token: String? {
        let path = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".config/vibewire/channel.token")
        return (try? String(contentsOf: path, encoding: .utf8))?
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func request(_ path: String, method: String = "GET") -> URLRequest? {
        guard let token, let url = URL(string: "http://127.0.0.1:\(port)\(path)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue(token, forHTTPHeaderField: "X-VibeWire-Channel")
        return request
    }

    /// Whether a session on this Mac currently has the channel loaded. The
    /// server only exists while Claude Code is running it, so this doubles as
    /// "is there a live session to join".
    func isAvailable() async -> Bool {
        guard var request = request("/health") else { return false }
        request.timeoutInterval = 1
        guard let (_, response) = try? await session.data(for: request) else { return false }
        return (response as? HTTPURLResponse)?.statusCode == 200
    }

    func setEmitter(_ emit: @escaping Emit) {
        self.emit = emit
    }

    /// Sends the phone's message into the live session.
    func send(text: String) async -> Bool {
        guard var request = request("/message", method: "POST") else { return false }
        request.httpBody = Data(text.utf8)
        guard let (data, response) = try? await session.data(for: request),
              (response as? HTTPURLResponse)?.statusCode == 200
        else {
            Log.warn(.claude, "channel send failed — is a session running with the channel loaded?")
            return false
        }
        if let body = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let chatId = body["chatId"] as? String {
            lastChatId = chatId
        }
        Log.info(.claude, "sent \(text.count) chars into the live session via the channel")
        return true
    }

    /// Answers a permission prompt the session raised. The terminal dialog stays
    /// open too, and Claude Code applies whichever verdict lands first.
    func answerPermission(requestId: String, allow: Bool) async {
        guard var request = request("/permission", method: "POST") else { return }
        request.httpBody = try? JSONSerialization.data(
            withJSONObject: ["requestId": requestId, "allow": allow]
        )
        _ = try? await session.data(for: request)
    }

    func attach() {
        guard streamTask == nil else { return }
        isAttached = true
        streamTask = Task { [weak self] in await self?.readEvents() }
    }

    func detach() {
        streamTask?.cancel()
        streamTask = nil
        isAttached = false
    }

    /// Reads the server-sent event stream: Claude's replies, and permission
    /// prompts raised by the live session.
    private func readEvents() async {
        guard let request = request("/events") else { return }
        do {
            let (stream, response) = try await session.bytes(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                Log.warn(.claude, "channel event stream refused")
                return
            }
            Log.info(.claude, "attached to the live session's channel")
            for try await line in stream.lines {
                guard !Task.isCancelled else { return }
                guard line.hasPrefix("data: ") else { continue }
                let payload = String(line.dropFirst(6))
                guard let data = payload.data(using: .utf8),
                      let event = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
                else { continue }
                handle(event)
            }
        } catch {
            guard !Task.isCancelled else { return }
            Log.warn(.claude, "channel event stream ended: \(error)")
        }
        isAttached = false
    }

    private func handle(_ event: [String: Any]) {
        switch event["kind"] as? String {
        case "reply":
            let text = event["text"] as? String ?? ""
            emit?([
                "t": "claude",
                "sub": "message",
                "role": "assistant",
                "blocks": [["type": "text", "text": text]],
            ])

        case "permission":
            emit?([
                "t": "claude",
                "sub": "permission",
                "requestId": event["requestId"] as? String ?? "",
                "toolName": event["toolName"] as? String ?? "",
                "command": event["inputPreview"] as? String ?? "",
                "explanation": event["description"] as? String ?? "",
                "waitingMs": 0,
            ])

        default:
            break
        }
    }
}

File: host-electron/src/main/claude/channelBridge.ts
import { readFileSync } from 'node:fs'
import { Log, describeError } from '../core/log'
import type { Payload } from '../net/wireProtocol'

/**
 * Talks to the VibeWire channel MCP server, which Claude Code spawns inside a
 * session that is already running. This is the difference between starting a
 * conversation from the phone and joining the one open in a terminal.
 *
 * The channel listens on loopback and is gated on a token, so this is the only
 * thing on the machine that can reach it.
 */
export class ChannelBridge {
  private emit: ((payload: Payload) => void) | null = null
  private abort: AbortController | null = null
  private attached = false

  constructor(
    private readonly tokenPath: string,
    private readonly port = 8790,
  ) {}

  private get token(): string | null {
    try {
      const token = readFileSync(this.tokenPath, 'utf8').trim()
      return token || null
    } catch {
      return null
    }
  }

  private headers(): Record<string, string> | null {
    const token = this.token
    return token ? { 'X-VibeWire-Channel': token } : null
  }

  private url(path: string): string {
    return `http://127.0.0.1:${this.port}${path}`
  }

  get isAttached(): boolean {
    return this.attached
  }

  setEmitter(emit: (payload: Payload) => void): void {
    this.emit = emit
  }

  /** Whether a session on this machine currently has the channel loaded. The
   *  server only exists while Claude Code is running it. */
  async isAvailable(): Promise<boolean> {
    const headers = this.headers()
    if (!headers) return false
    try {
      const response = await fetch(this.url('/health'), { headers, signal: AbortSignal.timeout(1000) })
      return response.status === 200
    } catch {
      return false
    }
  }

  /** Sends the phone's message into the live session. */
  async send(text: string): Promise<boolean> {
    const headers = this.headers()
    if (!headers) return false
    try {
      const response = await fetch(this.url('/message'), { method: 'POST', headers, body: text, signal: AbortSignal.timeout(10_000) })
      if (response.status !== 200) throw new Error(`status ${response.status}`)
      Log.info('claude', `sent ${text.length} chars into the live session via the channel`)
      return true
    } catch (error) {
      Log.warn('claude', `channel send failed — is a session running with the channel loaded? (${describeError(error)})`)
      return false
    }
  }

  /** Answers a permission prompt the session raised. The terminal dialog stays
   *  open too, and Claude Code applies whichever verdict lands first. */
  async answerPermission(requestId: string, allow: boolean): Promise<void> {
    const headers = this.headers()
    if (!headers) return
    try {
      await fetch(this.url('/permission'), {
        method: 'POST',
        headers: { ...headers, 'Content-Type': 'application/json' },
        body: JSON.stringify({ requestId, allow }),
        signal: AbortSignal.timeout(5000),
      })
    } catch (error) {
      Log.warn('claude', `channel permission answer failed: ${describeError(error)}`)
    }
  }

  attach(): void {
    if (this.abort) return
    this.attached = true
    this.abort = new AbortController()
    void this.readEvents(this.abort.signal)
  }

  detach(): void {
    this.abort?.abort()
    this.abort = null
    this.attached = false
  }

  /** Reads the server-sent event stream: Claude's replies, and permission
   *  prompts raised by the live session. */
  private async readEvents(signal: AbortSignal): Promise<void> {
    const headers = this.headers()
    if (!headers) return
    try {
      const response = await fetch(this.url('/events'), { headers, signal })
      if (response.status !== 200 || !response.body) {
        Log.warn('claude', 'channel event stream refused')
        return
      }
      Log.info('claude', "attached to the live session's channel")
      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''
      for (;;) {
        const { value, done } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })
        for (;;) {
          const newline = buffer.indexOf('\n')
          if (newline < 0) break
          const line = buffer.slice(0, newline).replace(/\r$/, '')
          buffer = buffer.slice(newline + 1)
          if (!line.startsWith('data: ')) continue
          try {
            this.handle(JSON.parse(line.slice(6)))
          } catch {
            // not JSON; skip
          }
        }
      }
    } catch (error) {
      if (!signal.aborted) Log.warn('claude', `channel event stream ended: ${describeError(error)}`)
    } finally {
      if (this.abort?.signal === signal) {
        this.abort = null
        this.attached = false
      }
    }
  }

  private handle(event: Record<string, unknown>): void {
    switch (event.kind) {
      case 'reply':
        this.emit?.({
          t: 'claude',
          sub: 'message',
          role: 'assistant',
          blocks: [{ type: 'text', text: typeof event.text === 'string' ? event.text : '' }],
        })
        break
      case 'permission':
        this.emit?.({
          t: 'claude',
          sub: 'permission',
          requestId: typeof event.requestId === 'string' ? event.requestId : '',
          toolName: typeof event.toolName === 'string' ? event.toolName : '',
          command: typeof event.inputPreview === 'string' ? event.inputPreview : '',
          explanation: typeof event.description === 'string' ? event.description : '',
          waitingMs: 0,
        })
        break
    }
  }
}
